src/lib/vasi-sync/src/lazy_lock.rs

```diff
@@ -0,0 +1,176 @@
+use core::marker::PhantomData;
+use core::mem::MaybeUninit;
+use core::sync::atomic::{self, AtomicI8};
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use vasi::VirtualAddressSpaceIndependent;
+
+use crate::sync::{ConstPtr, UnsafeCell};
+
+// For use with `LazyLock`. This trait is implemented for
+// `FnOnce` closures, so most users can just use that directly.
+// However in case they need to name a type in the LazyLock,
+// they can create a type that implements this trait instead.
+pub trait Producer<T> {
+    fn initialize(self) -> T;
+}
+
+impl<F, T> Producer<T> for F
+where
+    F: FnOnce() -> T,
+{
+    fn initialize(self) -> T {
+        (self)()
+    }
+}
+
+#[derive(TryFromPrimitive, IntoPrimitive, Eq, PartialEq)]
+#[repr(i8)]
+enum InitState {
+    Uninitd,
+    Initializing,
+    Initd,
+}
+
+/// Analogous to `std::sync::LazyLock`, but works on stable Rust, is no_std,
+/// only makes direct syscalls, etc.
+pub struct LazyLock<T, Init = fn() -> T> {
+    value: UnsafeCell<MaybeUninit<T>>,
+    init_state: AtomicI8,
+    initializer: UnsafeCell<Option<Init>>,
+}
+unsafe impl<T, Init> VirtualAddressSpaceIndependent for LazyLock<T, Init>
+where
+    T: VirtualAddressSpaceIndependent,
+    Init: VirtualAddressSpaceIndependent,
+{
+}
```

Comment:
Why implement it directly instead of using the derive? For example `Option` is not `VirtualAddressSpaceIndependent` and doesn't have a stable layout.